app/src/components/fields/CbSelectField.tsx

```diff
@@ -0,0 +1,144 @@
+} from '@mui/material';
+import { useFormikContext } from 'formik';
+import { ICbRouteKey, ICbSelectRows } from 'hooks/cb_api/useLookupApi';
+import { useCritterbaseApi } from 'hooks/useCritterbaseApi';
+import useDataLoader from 'hooks/useDataLoader';
+import get from 'lodash-es/get';
+import React, { ReactNode, useEffect } from 'react';
+
+interface ICbSelectSharedProps {
+  name: string;
+  label: string;
+  controlProps?: FormControlProps;
+}
+
+export interface ICbSelectField extends ICbSelectSharedProps {
+  id: string;
+  route: ICbRouteKey;
+  param?: string;
+  query?: string;
+  handleChangeSideEffect?: (value: string, label: string) => void;
+}
+
+interface ICbSelectOption {
+  value: string | number;
+  label: string;
+}
+/*
+ * Critterbase Select Field. Handles data retrieval, formatting and error handling.
+ *
+ * params {ICbSelectField}
+ * returns {*}
+ */
+
+const CbSelectField: React.FC<ICbSelectField> = (props) => {
+  const { name, label, route, param, query, handleChangeSideEffect, controlProps } = props;
+
+  const api = useCritterbaseApi();
+  const { data, load, isReady, refresh } = useDataLoader(api.lookup.getSelectOptions);
+  const { values, handleChange, setFieldValue, setFieldTouched } = useFormikContext<ICbSelectOption>();
+
+  const val = get(values, name) ?? '';
+
+  load({ route, param, query });
+
+  const isValueInRange = () => {
+    if (val === '') {
+      return true;
+    }
+    if (!data) {
+      return false;
+    }
+    return data.some((d) => (typeof d === 'string' ? d === val : d.id === val));
+  };
+
+  const handleInRange = () => {
+    if (!isValueInRange()) {
+      setFieldValue(name, '');
+      setFieldTouched(name);
+    }
+  };
+
+  useEffect(() => refresh({ route, param, query }), [route, param, query]);
+
+  useEffect(handleInRange, [isReady]);
+
+  const innerChangeHandler = (e: SelectChangeEvent<any>) => {
+    handleChange(e);
+    if (handleChangeSideEffect) {
+      const item = data?.find((a) => typeof a !== 'string' && a.id === e.target.value);
+      handleChangeSideEffect(e.target.value, (item as ICbSelectRows).value);
+    }
+  };
+
+  return (
+    <FormikSelectWrapper
+      name={name}
+      label={label}
+      controlProps={{ ...controlProps, disabled: !data?.length }}
+      onChange={innerChangeHandler}
+      value={isValueInRange() ? val : ''}>
+      {data?.map((a) => {
+        const item = typeof a === 'string' ? { label: a, value: a } : { label: a.value, value: a.id };
+        return (
+          <MenuItem key={item.value} value={item.value}>
+            {item.label}
+          </MenuItem>
+        );
+      })}
+    </FormikSelectWrapper>
+  );
+};
+
+interface FormikSelectWrapperProps extends ICbSelectSharedProps {
+  children?: ReactNode;
+  onChange?: SelectProps['onChange'];
+  value?: SelectProps['value'];
+}
+
+/**
+ *
+ * Wrapper for formik selects to handle all errors / onChange / onBlur
+ *
+ * Returns {*}
+ */
+
+export const FormikSelectWrapper = ({
```

Comment:
If this is a generic component, can we move it into its one file?

Reply:
Moved into fields directory / renamed to CbSelectFieldWrapper